### ISSUE TITLE
Surround and Back are reversed on AudioTrack

### ISSUE THREAD

**Issue by 618116** [xbmc/xbmc#22077]
### Describe the bug
Here is a clear and concise description of what the problem is:

When playing DTS-HD or True HD 7.1 ch on Android via NOT paththrough, surround channels and back channels are reversed.
For example, Dolby True HD have L/R/C/LFE/SL/SR/BL/BR layout, but Kodi output the audio as L/R/C/LFE/BL/BR/SL/SR.
Test tone is downloaded from https://www.demolandia.net/cinema/dolby-demo-trailers-hd/page-8.html

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

Surround and Back channels should be played as supposed to be.

## Actual Behavior

They are reversed.

## Possible Fix

I think this happen because CActiveAESink::OpenSink initialize AUDIOTRACK with channel layout FL, FR, FC, LFE, BL, BR, SL, SR.

### To Reproduce

1. Set every audio paththrough off.
2. Set number of audio channels as 7.1.
3. Play Dolby HD or DTS-HD 7.1 file.

### Debuglog

The debuglog can be found here:
[kodi.log](https://github.com/xbmc/xbmc/files/9900104/kodi.log)

### Screenshots
Here are some links or screenshots to help explain the problem:

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

### Your Environment
Used Operating system:

 - [x] Android
 - [ ] iOS
 - [ ] tvOS
 - [ ] Linux
 - [ ] OSX
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name:
 - Kodi version: 19.4 stable

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by fritsch** [comment]
We use:
https://developer.android.com/reference/android/media/AudioFormat#channelMask

especially: https://developer.android.com/reference/android/media/AudioFormat#CHANNEL_OUT_7POINT1 which is: 0x000003fc


public static final int CHANNEL_OUT_7POINT1
This constant was deprecated in API level 23.
Not the typical 7.1 surround configuration. Use [CHANNEL_OUT_7POINT1_SURROUND](https://developer.android.com/reference/android/media/AudioFormat#CHANNEL_OUT_7POINT1_SURROUND) instead.

S we should maybe use: which is CHANNEL_OUT_7POINT1_SURROUND 0x000018fc

Can you try that? When I give you a build?

**Comment by fritsch** [comment]
Sorry for the noise - nope, we use the proper "mask" - but that acutally might not do what we really want: https://github.com/xbmc/xbmc/blob/master/xbmc/cores/AudioEngine/Sinks/AESinkAUDIOTRACK.cpp#L145

Can your remove:
  if (info.Count() == 8 && info.HasChannel(AE_CH_LFE))
    return CJNIAudioFormat::CHANNEL_OUT_7POINT1_SURROUND;

Those two lines and rebuild? That should trigger:
https://github.com/xbmc/xbmc/blob/master/xbmc/cores/AudioEngine/Sinks/AESinkAUDIOTRACK.cpp#L149

which will manually assemble the channel mapping.

**Comment by fritsch** [comment]
The above linked PR will create a downloadable Android apk for v20 - please give this a try, remember do a backup of your kodi settings before as v20 will overwrite / forward port v19 settings.

**Comment by fritsch** [comment]
Here we go: https://jenkins.kodi.tv/job/Android-ARM/27105/artifact/kodi-20221031-78f67e4c-PR22078-27105-armeabi-v7a.apk

You seem to run 32-bit ARM build on your Shield TV. Not sure why - but therefore I linked you the 32 bit build.

**Comment by 618116** [comment]
> Here we go: https://jenkins.kodi.tv/job/Android-ARM/27105/artifact/kodi-20221031-78f67e4c-PR22078-27105-armeabi-v7a.apk
> 
> You seem to run 32-bit ARM build on your Shield TV. Not sure why - but therefore I linked you the 32 bit build.

Thank you for response!
Because I am using non-pro Shield(2019) which have 32bit OS. I'll try the apk when I'm home.

**Comment by 618116** [comment]
Unfortunately the problem is not fixed. Log file is attached.
L524 of log still shows wrong channel layout.
> 2022-11-01 23:37:38.642 T:28356   debug <general>: CActiveAESink::OpenSink - AUDIOTRACK Initialized:
> 2022-11-01 23:37:38.642 T:28356   debug <general>:   Output Device : AudioTrack (IEC)
> 2022-11-01 23:37:38.642 T:28356   debug <general>:   Sample Rate   : 48000
> 2022-11-01 23:37:38.642 T:28356   debug <general>:   Sample Format : AE_FMT_S16NE
> 2022-11-01 23:37:38.642 T:28356   debug <general>:   Channel Count : 8
> 2022-11-01 23:37:38.642 T:28356   debug <general>:   Channel Layout: FL, FR, FC, LFE, BL, BR, SL, SR
> 2022-11-01 23:37:38.643 T:28356   debug <general>:   Frames        : 2695
> 2022-11-01 23:37:38.643 T:28356   debug <general>:   Frame Size    : 16

[kodi.log](https://github.com/xbmc/xbmc/files/9910976/kodi.log)

**Comment by fritsch** [comment]
Good was worth a try.
Also good to see that the manual mixing results in the same pattern. AESinkAUDIOTRACK - channelMask 0x0018fc which is Android's default format for 7_1.

Now I remember a discussion about this two or three years ago - and in fact, the pattern was that it works fine for other devices besides the Shield

![image](https://user-images.githubusercontent.com/1089309/199281362-e5756a1b-6a66-4bfc-b4c6-fa8a15eabb8f.png)

If you look at the image ^^ - does this match your speaker positions for SL / BL?

**Comment by fritsch** [comment]
From your log, btw:
info <general>: ffmpeg[0x61955dd0]:   Stream #0:1(eng): Audio: truehd, 48000 Hz, 7.1, s32 (24 bit) (default)
7.1            FL+FR+FC+LFE+BL+BR+SL+SR
5.1            FL+FR+FC+LFE+BL+BR
5.1(side)      FL+FR+FC+LFE+SL+SR

Which is exactly the mapping kodi uses and also tells the Shield - so from this investigation.

See also the standard here:
https://xiph.org/flac/format.html
8 channels: front left, front right, front center, LFE, back left, back right, side left, side right

So - if you have another Android device to cross-check, which is not the Shield, please do so.

**Comment by 618116** [comment]
I have tried Fire TV cube(2022) and have same problem.
Also my speaker setup get correct result when play test tones from Windows and Xbox.

I think channel layout from this part,

> 2022-11-01 23:35:44.092 T:28355    info <general>: Enumerated AUDIOTRACK devices:
> 2022-11-01 23:35:44.092 T:28355    info <general>:     Device 1
> 2022-11-01 23:35:44.093 T:28355    info <general>:         m_deviceName      : AudioTrack (IEC)
> 2022-11-01 23:35:44.093 T:28355    info <general>:         m_displayName     : AudioTrack (IEC)
> 2022-11-01 23:35:44.093 T:28355    info <general>:         m_displayNameExtra: Kodi IEC packer (recommended)
> 2022-11-01 23:35:44.093 T:28355    info <general>:         m_deviceType      : AE_DEVTYPE_HDMI
> 2022-11-01 23:35:44.093 T:28355    info <general>:         m_channels        : FL, FR, FC, LFE, SL, SR, BL, BR, BC, BLOC, BROC

and this part,

> 2022-11-01 23:37:38.642 T:28356    info <general>: CAESinkAUDIOTRACK::Initializing with: m_sampleRate: 48000 format: AE_FMT_S16LE (AE) method: PCM stream-type: PCM-STREAM min_buffer_size: 172480 m_frames: 2695 m_frameSize: 16 channels: 8
> 2022-11-01 23:37:38.642 T:28356   debug <general>: CActiveAESink::OpenSink - AUDIOTRACK Initialized:
> 2022-11-01 23:37:38.642 T:28356   debug <general>:   Output Device : AudioTrack (IEC)
> 2022-11-01 23:37:38.642 T:28356   debug <general>:   Sample Rate   : 48000
> 2022-11-01 23:37:38.642 T:28356   debug <general>:   Sample Format : AE_FMT_S16NE
> 2022-11-01 23:37:38.642 T:28356   debug <general>:   Channel Count : 8
> 2022-11-01 23:37:38.642 T:28356   debug <general>:   Channel Layout: FL, FR, FC, LFE, BL, BR, SL, SR

should be same, isn't it?

**Comment by fritsch** [comment]
Nope. For passthrough it takes an entire different code path. The speaker names have zero impact as the data is compressed, just the AVR picked it out from there.

Can you play the same sample on windows Kodi please? What mapping do you get?

**Comment by fritsch** [comment]
And for enumeration, we just add the entire channel map available and decide when user actually open as file.
